src/Persistence_matrix/include/gudhi/Persistence_matrix/base_matrix_with_column_compression.h

```diff
@@ -0,0 +1,678 @@
+   * 
+   * @param columnIndex @ref MatIdx index of the column.
+   * @return true If the column has value zero.
+   * @return false Otherwise.
+   */
+  bool is_zero_column(index columnIndex);
+
+  /**
+   * @brief Resets the matrix to an empty matrix.
+   * 
+   * @param operators Pointer to the field operators.
+   * @param cellConstructor Pointer to the cell factory.
+   */
+  void reset(Column_settings* colSettings) {
+    columnToRep_.clear_and_dispose(delete_disposer());
+    columnClasses_ = boost::disjoint_sets_with_storage<>();
+    repToColumn_.clear();
+    nextColumnIndex_ = 0;
+    colSettings_ = colSettings;
+  }
+
+  /**
+   * @brief Assign operator.
+   */
+  Base_matrix_with_column_compression& operator=(const Base_matrix_with_column_compression& other);
+  /**
+   * @brief Swap operator.
+   */
+  friend void swap(Base_matrix_with_column_compression& matrix1, Base_matrix_with_column_compression& matrix2) {
+    matrix1.columnToRep_.swap(matrix2.columnToRep_);
+    std::swap(matrix1.columnClasses_, matrix2.columnClasses_);
+    matrix1.repToColumn_.swap(matrix2.repToColumn_);  // be careful when columnPool_ becomes not static
+    std::swap(matrix1.nextColumnIndex_, matrix2.nextColumnIndex_);
+    std::swap(matrix1.colSettings_, matrix2.colSettings_);
+
+    if constexpr (Master_matrix::Option_list::has_row_access) {
+      swap(static_cast<typename Master_matrix::Matrix_row_access_option&>(matrix1),
+           static_cast<typename Master_matrix::Matrix_row_access_option&>(matrix2));
+    }
+  }
+
+  void print();  // for debug
+
+ private:
+  /**
+   * @brief The disposer object function for boost intrusive container
+   */
+  struct delete_disposer {
+    void operator()(Column_type* delete_this) { columnPool_.destroy(delete_this); }
+  };
+
+  using ra_opt = typename Master_matrix::Matrix_row_access_option;
+  using col_dict_type = boost::intrusive::set<Column_type, boost::intrusive::constant_time_size<false> >;
+
+  col_dict_type columnToRep_;                         /**< Map from a column to the index of its representative. */
+  boost::disjoint_sets_with_storage<> columnClasses_; /**< Union-find structure,
+                                                           where two columns in the same set are identical. */
+  std::vector<Column_type*> repToColumn_;             /**< Map from the representative index to
+                                                           the representative Column. */
+  index nextColumnIndex_;                             /**< Next unused column index. */
+  Column_settings* colSettings_;                      /**< Cell factory. */
+  /**
+   * @brief Column factory.
+   * @warning As the member is static, they can eventually be problems if the matrix is duplicated in several threads.
+   * If this become necessary, the static should be removed in the future.
+   */
+  inline static Simple_object_pool<Column_type> columnPool_;
```

Comment:
It seems like a big problem if we cannot even have 2 threads that both compute persistence on completely independent filtrations.

Reply:
Ah, I forgot about that. I think, we can just remove the `static` from it and it should work just fine as it is only used inside the compressed matrix.

Reply:
True, but it is not that much of a problem. It was just to remind me that if I remove the static, I should also swap the pools in `swap` as the pointers won't belong to the right pool otherwise.

Reply:
So the right thing will happen if
- I swap 2 non-empty matrices
- I swap columns that don't belong to the same matrix

? Good then.

Reply:
If two non empty matrices are swapped, you just have to be careful that the matrices own the right pool at the end. If two columns from different matrices are swapped, their content are swapped but not their addresses, so it is also fine. Were problem can arise when swapping two columns from different matrices, is at the row access, as the cells in the rows will not be swapped (you should first unlink a column and relink it after swap). But this has nothing to do with this pool. Nevertheless, I realized that I never mentioned this in the doc. I could also just add the linking/unlinking in the swap operator.